### ISSUE TITLE
[Hold #50639] [$250]  Category - Long press a category&tap outside highlights the category in mWeb & not in hybrid

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#51394]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** V9. 0.53-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N
**Issue reported by:** Applause Internet Team

## Action Performed:
1. Launch site in both mweb and hybrid app
2. Tap profile icon -- Workspaces-- workspace
3. Tap category
4. Long press on a category and tap outside
5. Note the category is highlighted in mweb and not in hybrid app

## Expected Result:
Long pressing on a category and tapping outside must show same behavior in mweb and hybrid app.

## Actual Result:
Long pressing on a category and tapping outside highlights the category in mweb and not in hybrid app.
Issue reproduced only in redmi devices.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/c6d56db6-957b-422f-b2c8-b3b2a38b7dd2

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021853406887958563486</li>
        <li>Upwork Job ID: 1853406887958563486</li>
        <li>Last Price Increase: 2024-11-18</li>
    </ul>
</details>

**Comment by IuliiaHerets** [comment]
@trjExpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by trjExpensify** [comment]
>Issue reproduced only in redmi devices.

Given this, I'm inclined to close this issue. @Expensify/design for a quick hip check before I do.

**Comment by dubielzyk-expensify** [comment]
Yeah I can't reproduce this

**Comment by trjExpensify** [comment]
Great stuff, closing then.

**Comment by shahinyan11** [comment]
@trjExpensify @dubielzyk-expensify The problem is still reproducible.

https://github.com/user-attachments/assets/588d2dfe-10be-4e10-96f4-033564265f50

**Comment by trjExpensify** [comment]
@shahinyan11 on a Redmi device only?

**Comment by shahinyan11** [comment]
> @shahinyan11 on a Redmi device only?

On this simulator

![Screenshot 2024-11-01 at 14 10 31](https://github.com/user-attachments/assets/ad1fefec-3964-407d-bca3-a02f74f2acfb)

**Comment by trjExpensify** [comment]
Oh hm, @dubielzyk-expensify what Android device did you test?

**Comment by dubielzyk-expensify** [comment]
Tested it on a Pixel 7

**Comment by dubielzyk-expensify** [comment]
My bad. I can actually reproduce it now. I'm still not overly convinced we should fix it, but I'll leave that decision to you @trjExpensify

**Comment by trjExpensify** [comment]
Ah okay, so yeah.. I think we should at least investigate the root cause then if it's even reproducible on a Pixel.

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Category - Long press a category&tap outside highlights the category in mWeb & not in hybrid
### What is the root cause of that problem?
[Here](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L479C25-L481C26) we have a code to prevent focus and we prevent it during `onLongPress` action on mobile web platform ( [See here](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/index.tsx#L46C1-L48C76) ). But currently the `onFocus` event is also triggered after closing the modal that is opened by a long press action.
### What changes do you think we should make in order to solve the problem?
Move the prevent logic inside `BaseListItem` component and blur the  focused item before `return`

1. Add `shouldIgnoreFocus?: boolean` in [ListItem](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/types.ts#L93) type
2. Pass `shouldIgnoreFocus` as a prop to [ListItem](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L454C18-L454C26) component
3. Add `shouldIgnoreFocus` in `TableListItem` component's [props](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L15C48-L28C2) and pass following prop `shouldIgnoreFocus={shouldIgnoreFocus}`  to  [BaseListItem](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L45C10-L45C22)
4. Add `shouldIgnoreFocus` in `BaseListItem` component's [props](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L16C10-L16C22) and update this [onFocus](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L102C26-L102C33) to bellow
```
onFocus={()=>{
    if(shouldIgnoreFocus || isDisabled) {
        pressableRef.current?.blur()
        return;
    }
    onFocus()
}}
```


### What alternative solutions did you explore? (Optional)

**Comment by Jaeta01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Category - Long press a category&tap outside highlights the category in mWeb & not in hybrid
### What is the root cause of that problem?
In [this](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L479C25-L481C26) case, we have implemented code to prevent focus during an onLongPress action on the mobile web platform (as referenced [here](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/index.tsx#L46C1-L48C76)). However, the issue is that the `onFocus` event is still triggered after the modal—opened by the long press action—is closed.
### What changes do you think we should make in order to solve the problem?
Move the prevent logic inside BaseListItem component and blur the focused item before return

1. Pass shouldIgnoreFocus as a prop to [ListItem](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L454C18-L454C26) component. ( Also need to add `shouldIgnoreFocus?: boolean` value in [ListItem](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/types.ts#L93) type )
2. Add shouldIgnoreFocus in TableListItem component's [props](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L15C48-L28C2) and pass following prop shouldIgnoreFocus={shouldIgnoreFocus} to [BaseListItem](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L45C10-L45C22)
3. Add shouldIgnoreFocus in BaseListItem component's [props](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L16C10-L16C22) and update this [onFocus](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L102C26-L102C33) to bellow

```
onFocus={()=>{
    if(shouldIgnoreFocus || isDisabled) {
        pressableRef.current?.blur()
        return;
    }
    onFocus()
}}
```
### What alternative solutions did you explore? (Optional)

**Comment by c3024** [comment]
I will check the proposal and update.

**Comment by c3024** [comment]
Waiting for [valid](https://expensify.slack.com/archives/C02NK2DQWUX/p1731577221526339?thread_ts=1713200389.937059&cid=C02NK2DQWUX) proposals!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by QichenZhu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

After long-pressing a selection list item and dismissing the context menu, the item is highlighted.

### What is the root cause of that problem?

The item is focused by long-pressing, but the conditions below detect that the focus is triggered by touch and ignore it, so the item is not highlighted immediately.

https://github.com/Expensify/App/blob/e4969b22e7dbb1f6a200cc587ced4ab665415c0e/src/components/SelectionList/index.tsx#L48

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/components/SelectionList/BaseSelectionListItemRenderer.tsx#L77-L79

Instead, it is highlighted only after the context menu is closed, when the focus trap programmatically returns focus to the item.

https://github.com/focus-trap/focus-trap/blob/6c757fa2a90bb818b7f702ba4129d14cdc733f7e/index.js#L769

<img width="601" alt="Screenshot 2024-11-13 at 2 51 43 PM" src="https://github.com/user-attachments/assets/6f91f33c-4fad-442e-834f-ccddb6c6e902">

### What changes do you think we should make in order to solve the problem?
In addition to ignoring focus triggered by touch, we should also ignore programmatic focus.

```TSX
import * as Browser from '@libs/Browser';
```

```TSX
onFocus={(event) => {
```
https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/components/SelectionList/BaseSelectionListItemRenderer.tsx#L77-L79

```TSX
    // Insert this after the above code.
    if (Browser.isMobileChrome() && event && event.nativeEvent && !event.nativeEvent.sourceCapabilities) {
        return;
    }
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by wildan-m** [comment]
It has the same root cause with https://github.com/Expensify/App/issues/50639

My proposal for the issue: https://github.com/Expensify/App/issues/50639#issuecomment-2456750349

**Comment by gev1990** [comment]
## Proposal
Note I have noticed a proposal here from another member that does not currently exist, and since I have liked  the main concept of that proposal I decided to post a new proposal that may be similar to the proposal that was there previously.

### Please re-state the problem that we are trying to solve in this issue.
Category - Long press a category&tap outside highlights the category in mWeb & not in hybrid

### What is the root cause of that problem?
[Here](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L479C25-L481C26) we ignore focus when `shouldIgnoreFocus` is `true`. And [here](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/index.tsx#L46C1-L48C76) you can see a description in which cases `shouldIgnoreFocus` will be `true`
Now  the problem occurs because [onFocus](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L478C1-L483C23) event is also triggered after closing modal when the `shouldIgnoreFocus` is `false`.

### What changes do you think we should make in order to solve the problem?
Move the ignore focus logic inside `BaseListItem` component and blur the focused item before return

1. Pass `shouldIgnoreFocus` as a prop to [ListItem](https://github.com/Expensify/App/blob/4aa007bf405bd11b275bf3e599817e192fa2b249/src/components/SelectionList/BaseSelectionList.tsx#L454C18-L454C26) component.
2. Add `shouldIgnoreFocus` in [TableListItem](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L15) component's [props](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L15C48-L28C2) and pass following prop `shouldIgnoreFocus={shouldIgnoreFocus}` to [BaseListItem](https://github.com/Expensify/App/blob/e1e970d27f3c40a4cd6cfc80d02b9d4c8619aa31/src/components/SelectionList/TableListItem.tsx#L45C10-L45C22)
3. Add `shouldIgnoreFocus` in `BaseListItem` component's [props](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L16C10-L16C22) and update this [onFocus](https://github.com/Expensify/App/blob/945421dfdd575a98b4cea49e5385453339ce800a/src/components/SelectionList/BaseListItem.tsx#L102C26-L102C33) to bellow
```
onFocus={()=>{
    if(shouldIgnoreFocus || isDisabled) {
        pressableRef.current?.blur()
        return;
    }
    onFocus()
}}
```

### What alternative solutions did you explore? (Optional)

**Comment by trjExpensify** [comment]
> It has the same root cause with #50639
> 
> My proposal for the issue: [#50639 (comment)](https://github.com/Expensify/App/issues/50639#issuecomment-2456750349)

Given that, @c3024 do you think this is a dupe?

**Comment by c3024** [comment]
Yes, I think fixing that issue should fix this too. We can hold off this issue for #50639.

**Comment by trjExpensify** [comment]
Sounds good to me, putting it on hold.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by trjExpensify** [comment]
Still held on this PR: https://github.com/Expensify/App/pull/53120

**Comment by trjExpensify** [comment]
Same as above, Melv.

**Comment by trjExpensify** [comment]
Same as above, Melv.

**Comment by trjExpensify** [comment]
Same, melv.

**Comment by trjExpensify** [comment]
Bumped the PR we're waiting on.

**Comment by trjExpensify** [comment]
@c3024 https://github.com/Expensify/App/pull/53120 has been deployed, do you want to check this issue now?

**Comment by c3024** [comment]
@trjExpensify , this was fixed. This is not reproducible anymore. We can close this.

https://github.com/user-attachments/assets/df29196b-0099-4755-a5eb-3abcc0c8cd7c

**Comment by trjExpensify** [comment]
Dope, closing!
